File: controllers/lessonController.js
const prisma = require('../prismaClient');

// Add a lesson to a course
exports.addLesson = async (req, res) => {
  console.log("req.params", req.params);
  console.log("req.body", req.body);
  const { courseId } = req.params;
  const { title, description, videoUrl, notes, orderIndex } = req.body;

  try {
    const lesson = await prisma.lesson.create({
      data: {
        title,
        description,
        videoUrl,
        orderIndex: parseInt(orderIndex),
        course: {
          connect: { id: parseInt(courseId) },
        },
        
      }
    });
    res.status(201).json(lesson);

  } catch (error) {
    console.log(error);
    res.status(500).json({ error: 'Failed to add lesson', details: error.message });
  }
};

// Get all lessons of a course
exports.getCourseLessons = async (req, res) => {
  const { courseId } = req.params;

  try {
    const lessons = await prisma.lesson.findMany({
      where: { courseId: parseInt(courseId) },
      orderBy: { orderIndex: 'asc' }
    });
    res.status(200).json(lessons);
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch lessons', details: error.message });
  }
};

// Update a lesson by ID
exports.updateLesson = async (req, res) => {
  const { lessonId } = req.params;
  const { title, description, videoUrl, notes, orderIndex } = req.body;

  try {
    const updatedLesson = await prisma.lesson.update({
      where: { id: parseInt(lessonId) },
      data: {
        title,
        description,
        videoUrl,
        notes,
        orderIndex: parseInt(orderIndex),
      },
    });
    res.status(200).json(updatedLesson);
  } catch (error) {
    res.status(500).json({ error: 'Failed to update lesson', details: error.message });
  }
};

// Delete a lesson by ID
exports.deleteLesson = async (req, res) => {
  const { lessonId } = req.params;

  try {
    await prisma.lesson.delete({
      where: { id: parseInt(lessonId) },
    });
    res.status(200).json({ message: 'Lesson deleted successfully' });
  } catch (error) {
    res.status(500).json({ error: 'Failed to delete lesson', details: error.message });
  }
};

// **NEW: Bulk update lessons for a course**
exports.bulkUpdateLessons = async (req, res) => {
  const { courseId } = req.params;
  const { lessons } = req.body;

  if (!lessons || !Array.isArray(lessons)) {
    return res.status(400).json({ error: 'Lessons array is required' });
  }

  try {
    // Use transaction to ensure data consistency
    const result = await prisma.$transaction(async (tx) => {
      // Get existing lessons for this course
      const existingLessons = await tx.lesson.findMany({
        where: { courseId: parseInt(courseId) },
        select: { id: true, title: true, description: true, videoUrl: true, notes: true, orderIndex: true }
      });

      const existingLessonMap = new Map(existingLessons.map(lesson => [lesson.id, lesson]));
      const incomingLessonIds = new Set(lessons.filter(l => l.id).map(l => parseInt(l.id)));
      
      const operations = {
        created: [],
        updated: [],
        deleted: []
      };

      // Process incoming lessons
      for (const lesson of lessons) {
        const lessonData = {
          title: lesson.title,
          description: lesson.description,
          videoUrl: lesson.videoUrl,
          notes: lesson.notes,
          orderIndex: parseInt(lesson.orderIndex)
        };

        if (lesson.id) {
          // Update existing lesson
          const lessonId = parseInt(lesson.id);
          if (existingLessonMap.has(lessonId)) {
            const updatedLesson = await tx.lesson.update({
              where: { id: lessonId },
              data: lessonData
            });
            operations.updated.push(updatedLesson);
          }
        } else {
          // Create new lesson
          const newLesson = await tx.lesson.create({
            data: {
              ...lessonData,
              course: {
                connect: { id: parseInt(courseId) }
              }
            }
          });
          operations.created.push(newLesson);
        }
      }

      // Delete lessons that are no longer present
      const lessonsToDelete = existingLessons.filter(lesson => !incomingLessonIds.has(lesson.id));
      
      for (const lesson of lessonsToDelete) {
        await tx.lesson.delete({
          where: { id: lesson.id }
        });
        operations.deleted.push({ id: lesson.id, title: lesson.title });
      }

      return operations;
    });

    res.status(200).json({
      message: 'Lessons updated successfully',
      summary: {
        created: result.created.length,
        updated: result.updated.length,
        deleted: result.deleted.length
      },
      details: result
    });

  } catch (error) {
    console.log(error);
    res.status(500).json({ 
      error: 'Failed to bulk update lessons', 
      details: error.message 
    });
  }
};

// Mark Lesson as completed
exports.markLessonCompleted = async (req, res) => {
  const { lessonId, enrollmentId } = req.params;
  
  try {
    const lessonCompletion = await prisma.lessonCompletion.upsert({
      where: {
        enrollmentId_lessonId: {
          lessonId: parseInt(lessonId, 10),
          enrollmentId: parseInt(enrollmentId, 10)
        }
      },
      update: {},
      create: {
        lessonId: parseInt(lessonId, 10),
        enrollmentId: parseInt(enrollmentId, 10)
      }
    });
    
    // If we're updating an existing record, it means it was already marked as completed
    const isNew = lessonCompletion.createdAt === lessonCompletion.updatedAt;
    
    res.status(200).json({
      ...lessonCompletion,
      isNew
    });
  } catch (error) {
    console.log(error);
    res.status(500).json({ error: 'Failed to mark lesson as completed', details: error.message });
  }
};

// API Endpoints:
// create lesson
// POST -> http://localhost:5000/api/lessons/add/:courseId
// request body: {
//   "title": "Lesson 1",
//   "description": "This is a lesson description.",
//   "videoUrl": "http://localhost:5000/api/lessons/video",
//   "notes": "This is a lesson notes.",
//   "orderIndex": 0
// }

// update lesson
// PUT -> http://localhost:5000/api/lessons/update/:lessonId
// request body: {
//   "title": "Introduction to programming"
// }

// delete lesson
// DELETE -> http://localhost:5000/api/lessons/delete/:lessonId

// get all lessons of a course
// GET -> http://localhost:5000/api/lessons/get/:courseId

// **NEW: bulk update lessons**
// PUT -> http://localhost:5000/api/lessons/bulk/:courseId
// request body: {
//   "lessons": [
//     {
//       "id": 1, // existing lesson - will be updated
//       "title": "Updated Lesson 1",
//       "description": "Updated description",
//       "videoUrl": "http://example.com/video1",
//       "notes": "Updated notes",
//       "orderIndex": 0
//     },
//     {
//       // no id - will be created as new lesson
//       "title": "New Lesson 2",
//       "description": "New lesson description",
//       "videoUrl": "http://example.com/video2",
//       "notes": "New lesson notes",
//       "orderIndex": 1
//     }
//     // lessons not in this array but existing in DB will be deleted
//   ]
// }

// mark lesson as completed
// POST -> http://localhost:5000/api/lessons/mark-completed/:lessonId/:enrollmentId

File: src/main/java/com/eduverse/courseservice/controller/LessonController.java
package com.eduverse.courseservice.controller;

import java.util.List;
import java.util.Map;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.eduverse.courseservice.dto.BulkLessonsRequest;
import com.eduverse.courseservice.dto.BulkLessonsResponse;
import com.eduverse.courseservice.dto.LessonDto;
import com.eduverse.courseservice.service.LessonService;

@RestController
@RequestMapping("/api/lessons")
@CrossOrigin(origins = "${cors.allowed-origins}")
public class LessonController {

    private final LessonService lessonService;

    public LessonController(LessonService lessonService) {
        this.lessonService = lessonService;
    }

    // Add a lesson to a course
    @PostMapping("/add/{courseId}")
    public ResponseEntity<Object> addLesson(@PathVariable Long courseId, @RequestBody Map<String, Object> request) {
        try {
            String title = (String) request.get("title");
            String description = (String) request.get("description");
            String videoUrl = (String) request.get("videoUrl");
            String notes = (String) request.get("notes");
            Integer orderIndex = (Integer) request.get("orderIndex");
            
            LessonDto lessonDto = new LessonDto();
            lessonDto.setTitle(title);
            lessonDto.setDescription(description);
            lessonDto.setVideoUrl(videoUrl);
            lessonDto.setNotes(notes);
            lessonDto.setOrderIndex(orderIndex);
            
            LessonDto createdLesson = lessonService.addLesson(courseId, lessonDto);
            return ResponseEntity.status(HttpStatus.CREATED).body(createdLesson);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Map.of(
                "error", "Failed to add lesson",
                "details", e.getMessage()
            ));
        }
    }

    // Get all lessons of a course
    @GetMapping("/get/{courseId}")
    public ResponseEntity<Object> getCourseLessons(@PathVariable Long courseId) {
        try {
            List<LessonDto> lessons = lessonService.getCourseLessons(courseId);
            return ResponseEntity.ok(lessons);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Map.of(
                "error", "Failed to fetch lessons",
                "details", e.getMessage()
            ));
        }
    }

    // Get lesson by ID
    @GetMapping("/get/lesson/{id}")
    public ResponseEntity<LessonDto> getLessonById(@PathVariable Long id) {
        LessonDto lesson = lessonService.getLessonById(id);
        return ResponseEntity.ok(lesson);
    }

    // Update a lesson by ID
    @PutMapping("/update/{lessonId}")
    public ResponseEntity<Object> updateLesson(@PathVariable Long lessonId, @RequestBody Map<String, Object> request) {
        try {
            String title = (String) request.get("title");
            String description = (String) request.get("description");
            String videoUrl = (String) request.get("videoUrl");
            String notes = (String) request.get("notes");
            Integer orderIndex = (Integer) request.get("orderIndex");
            
            LessonDto lessonDto = new LessonDto();
            lessonDto.setTitle(title);
            lessonDto.setDescription(description);
            lessonDto.setVideoUrl(videoUrl);
            lessonDto.setNotes(notes);
            lessonDto.setOrderIndex(orderIndex);
            
            LessonDto updatedLesson = lessonService.updateLesson(lessonId, lessonDto);
            return ResponseEntity.ok(updatedLesson);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Map.of(
                "error", "Failed to update lesson",
                "details", e.getMessage()
            ));
        }
    }

    // Delete a lesson by ID
    @DeleteMapping("/delete/{lessonId}")
    public ResponseEntity<Object> deleteLesson(@PathVariable Long lessonId) {
        try {
            lessonService.deleteLesson(lessonId);
            return ResponseEntity.ok(Map.of("message", "Lesson deleted successfully"));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Map.of(
                "error", "Failed to delete lesson",
                "details", e.getMessage()
            ));
        }
    }

    // Bulk update lessons for a course
    @PutMapping("/bulk/{courseId}")
    public ResponseEntity<Object> bulkUpdateLessons(
            @PathVariable Long courseId, 
            @RequestBody Map<String, Object> request) {
        try {
            @SuppressWarnings("unchecked")
            List<Map<String, Object>> lessons = (List<Map<String, Object>>) request.get("lessons");
            
            if (lessons == null) {
                return ResponseEntity.badRequest().body(Map.of("error", "Lessons array is required"));
            }
            
            BulkLessonsRequest bulkRequest = new BulkLessonsRequest();
            bulkRequest.setLessons(lessons.stream().map(lessonMap -> {
                LessonDto lessonDto = new LessonDto();
                if (lessonMap.get("id") != null) {
                    lessonDto.setId(Long.valueOf(lessonMap.get("id").toString()));
                }
                lessonDto.setTitle((String) lessonMap.get("title"));
                lessonDto.setDescription((String) lessonMap.get("description"));
                lessonDto.setVideoUrl((String) lessonMap.get("videoUrl"));
                lessonDto.setNotes((String) lessonMap.get("notes"));
                if (lessonMap.get("orderIndex") != null) {
                    lessonDto.setOrderIndex(Integer.valueOf(lessonMap.get("orderIndex").toString()));
                }
                return lessonDto;
            }).collect(java.util.stream.Collectors.toList()));
            
            BulkLessonsResponse response = lessonService.bulkUpdateLessons(courseId, bulkRequest);
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Map.of(
                "error", "Failed to bulk update lessons",
                "details", e.getMessage()
            ));
        }
    }

    // Mark lesson as completed
    @PostMapping("/mark-completed/{lessonId}/{enrollmentId}")
    public ResponseEntity<Object> markLessonCompleted(
            @PathVariable Long lessonId, 
            @PathVariable Long enrollmentId) {
        try {
            Map<String, Object> response = lessonService.markLessonCompleted(lessonId, enrollmentId);
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Map.of(
                "error", "Failed to mark lesson as completed",
                "details", e.getMessage()
            ));
        }
    }
}
